File: irrgarten-java/src/irrgarten/Labyrinth.java
package irrgarten;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Labyrinth {

    private static final char BLOCK_CHAR = 'X';
    private static final char EMPTY_CHAR = '-';
    private static final char MONSTER_CHAR = 'M';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;

    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;

    private Monster[][] monsters;
    private char[][] labyrinth;
    private Player[][] players;

    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol) {
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;

        monsters = new Monster[nRows][nCols];
        labyrinth = new char[nRows][nCols];
        players = new Player[nRows][nCols];

        for (int i = 0; i < nRows; i++) {
            Arrays.fill(labyrinth[i], EMPTY_CHAR);
        }

        labyrinth[exitRow][exitCol] = EXIT_CHAR;
    }

    public void spreadPlayers(Player[] players) {
        for (Player player : players) {
            int[] pos = randomEmptyPos();
            putPlayer2D(-1, -1, pos[ROW], pos[COL], player);       
        }
    }

    public boolean haveAWinner() {
        return players[exitRow][exitCol] != null;
    }

    // TODO: creo que aqui hay que imprimirlo como tabla
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nCols; j++) {
                sb.append(labyrinth[i][j]);
                sb.append(" ");
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    public void addMonster(int row, int col, Monster monster) {
        if (posOK(row, col) && emptyPos(row, col)) {
            monsters[row][col] = monster;
            monster.setPos(row, col);
        }
    }

    public Monster putPlayer(Directions direction, Player player) {
        int oldRow = player.getRow();
        int oldCol = player.getCol();
        int[] newPos = dir2Pos(oldRow, oldCol, direction);
        return putPlayer2D(oldRow, oldCol, newPos[ROW], newPos[COL], player);
    }

    public void addBlock(Orientation orientation, int startRow, int startCol, int length) {
        int incRow, incCol;
        if (orientation == Orientation.VERTICAL) {
            incRow = 1;
            incCol = 0;
        } else {
            incRow = 0;
            incCol = 1;
        }
        
        int row = startRow;
        int col = startCol;
        
        while (posOK(row, col) && emptyPos(row, col) && length > 0) {
            labyrinth[row][col] = BLOCK_CHAR;
            length--;
            row += incRow;
            col += incCol;
        }
    }

    public Directions[] validMoves(int row, int col) {
        List<Directions> output = new ArrayList<>();
        if (canStepOn(row+1, col)) {
            output.add(Directions.DOWN);
        }
        if (canStepOn(row-1, col)) {
            output.add(Directions.UP);
        }
        if (canStepOn(row, col+1)) {
            output.add(Directions.RIGHT);
        }
        if (canStepOn(row, col-1)) {
            output.add(Directions.LEFT);
        }
        
        
        return output.toArray(new Directions[output.size()]);
    }

    private boolean posOK(int row, int col) {
        return 0 <= row && 0 <= col && nRows > row && nCols > col;
    }

    private boolean emptyPos(int row, int col) {
        return labyrinth[row][col] == EMPTY_CHAR;
    }

    private boolean monsterPos(int row, int col) {
        return labyrinth[row][col] == MONSTER_CHAR;
    }

    private boolean exitPos(int row, int col) {
        return labyrinth[row][col] == EXIT_CHAR;
    }

    private boolean combatPos(int row, int col) {
        return labyrinth[row][col] == COMBAT_CHAR;
    }

    private boolean canStepOn(int row, int col) {
        return posOK(row, col)
                && !combatPos(row, col)
                && labyrinth[row][col] != BLOCK_CHAR;
    }

    private void updateOldPos(int row, int col) {
        if (posOK(row, col)) {
            labyrinth[row][col] = combatPos(row, col) ? MONSTER_CHAR : EMPTY_CHAR;
        }
    }

    private int[] dir2Pos(int row, int col, Directions direction) {
        int[] out = {row, col};
        
        switch (direction) {
            case DOWN  -> out[ROW]++;
            case UP    -> out[ROW]--;
            case LEFT  -> out[COL]--;
            case RIGHT -> out[COL]++;
        }
                
        return out;
    }
    
    private int[] randomEmptyPos() {
        int row, col;
        do {
            row = Dice.randomPos(nRows);
            col = Dice.randomPos(nCols);
        } while (!emptyPos(row, col));
        return new int[]{row, col};
    }
    
    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player) {
        Monster output = null;
        if (canStepOn(row, col)) {
            if (posOK(oldRow, oldCol)) {
                Player p = players[oldRow][oldCol];
                if (p == player) {
                    updateOldPos(oldRow, oldCol);
                    players[oldRow][oldCol] = null;
                }
            }
            
            if (monsterPos(row, col)) {
                labyrinth[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            } else {
                labyrinth[row][col] = player.getNumber();
            }
            
            players[row][col] = player;
            player.setPos(row, col);
            
        }
        return output;
    }
}

File: irrgarten-ruby/labyrinth.rb
# encoding: utf-8
require_relative 'directions'

module Irrgarten
  class Labyrinth
    @@BLOCK_CHAR = 'X'
    @@EMPTY_CHAR = '-'
    @@MONSTER_CHAR = 'M'
    @@COMBAT_CHAR = 'C'
    @@EXIT_CHAR = 'E'
    @@ROW = 0
    @@COL = 1

    def initialize(nRows, nCols, exitRow, exitCol)
      @n_rows = nRows
      @n_cols = nCols
      @exit_row = exitRow
      @exit_col = exitCol

      @players = Array.new(@n_rows) { Array.new(@n_cols) }
      @monsters = Array.new(@n_rows) { Array.new(@n_cols) }
      @labyrinth = Array.new(@n_rows) { Array.new(@n_cols, @@EMPTY_CHAR) }

      @labyrinth[@exit_row][@exit_col] = @@EXIT_CHAR
    end

    def spread_players(players)
      players.each do |player|
        pos = random_empty_pos
        put_player_2d(-1, -1, pos[@@ROW], pos[@@COL], player)
      end
    end

    def have_a_winner
      @players[@exit_row][@exit_col] != nil
    end

    def to_s
      str = ""

      (0...@n_rows).each do |i|
        (0...@n_cols).each do |j|
          str += @labyrinth[i][j].to_s + " "
        end
        str += "\n"
      end

      str
    end

    def add_monster(row, col, monster)
      if pos_ok(row, col) && empty_pos(row, col)
        @monsters[row][col] = monster
        monster.set_pos(row, col)
      end
    end

    def put_player(direction, player)
      old_row = player.row
      old_col = player.col
      new_pos = dir_2_pos(old_row, old_col, direction)
      put_player_2d(old_row, old_col, new_pos[@@ROW], new_pos[@@COL], player)
    end

    def add_block(orientation, start_row, start_col, length)
      inc_row = 0
      inc_col = 1
      if orientation == Orientation::VERTICAL
        inc_row = 1
        inc_col = 0
      end
      row = start_row
      col = start_col

      while pos_ok(row, col) && empty_pos(row, col) && length > 0
        @labyrinth[row][col] = @@BLOCK_CHAR
        length -= 1
        row += inc_row
        col += inc_col
      end
    end

    def valid_moves(row, col)
      output = []
      if can_step_on(row + 1, col)
        output.push(Directions::DOWN)
      end
      if can_step_on(row - 1, col)
        output.push(Directions::UP)
      end
      if can_step_on(row, col + 1)
        output.push(Directions::RIGHT)
      end
      if can_step_on(row, col - 1)
        output.push(Directions::LEFT)
      end

      output
    end

    private

    def pos_ok(row, col)
      row.between?(0, @n_rows - 1) && col.between?(0, @n_cols - 1)
    end

    def empty_pos(row, col)
      @labyrinth[row][col] == @@EMPTY_CHAR
    end

    def monster_pos(row, col)
      @labyrinth[row][col] == @@MONSTER_CHAR
    end

    def exit_pos(row, col)
      @labyrinth[row][col] == @@EXIT_CHAR
    end

    def combat_pos(row, col)
      @labyrinth[row][col] == @@COMBAT_CHAR
    end

    def can_step_on(row, col)
      pos_ok(row, col) && !combat_pos(row, col) && @labyrinth[row][col] != @@BLOCK_CHAR
    end

    def update_old_pos(row, col)
      if pos_ok(row, col)
        @labyrinth[row][col] = combat_pos(row, col) ? @@MONSTER_CHAR : @@EMPTY_CHAR
      end
    end

    def dir_2_pos(row, col, directions)
      out = [row, col]
      case directions
      when Directions::DOWN
        out[@@ROW] += 1
      when Directions::UP
        out[@@ROW] -= 1
      when Directions::LEFT
        out[@@COL] -= 1
      when Directions::RIGHT
        out[@@COL] += 1
      end
      out
    end

    def random_empty_pos
      row = 0
      col = 0
      loop do
        row = Dice.random_pos(@n_rows)
        col = Dice.random_pos(@n_cols)
        break if empty_pos(row, col)
      end
      [row, col]
    end

    def put_player_2d(old_row, old_col, row, col, player)
      output = nil
      if can_step_on(row, col)
        if pos_ok(old_row, old_col) && @players[old_row][old_col] == player
          update_old_pos(old_row, old_col)
          @players[old_row][old_col] = nil
        end

        if monster_pos(row, col)
          @labyrinth[row][col] = @@COMBAT_CHAR;
          output = @monsters
        else
          @labyrinth[row][col] = player.number.to_s
        end
        @players[row][col] = player
        player.set_pos(row, col)
      end

      output
    end
  end
end

